include/onnxruntime/core/common/gpu_profiler_common.h

```diff
@@ -0,0 +1,397 @@
+#pragma once
+
+#include "core/common/profiler_common.h"
+#include "core/common/inlined_containers.h"
+
+#include <map>
+#include <memory>
+#include <mutex>
+#include <sstream>
+#include <string>
+#include <vector>
+#include <utility>
+
+namespace onnxruntime {
+namespace profiling {
+
+// The classes in this header are implemented as template/inline classes
+// to avoid having to export symbols from the main onnxruntime shared library
+// to ExecutionProvider (EP) shared libraries.
+// More context: The main onnxruntime shared library is optimized for size
+// using --gc-sections during link time to ensure that any unreferenced code
+// is not retained. This poses a problem in using a design pattern where the
+// (abstract) base class is implemented in the main onnxruntime shared library,
+// but (concrete) subclasses are implemented in EP shared libraries. Now, because
+// EP shared libraries are loaded at runtime (as of 11/2022), there will be no
+// references to the base class symbols when the main onnxruntime shared library
+// is compiled. Thus, the base class symbols will not be included in the
+// main onnxruntime shared library. This manifests in being unable to load
+// EP shared libs (because the base class symbols referenced by derived
+// classes are missing).
+// We solve this by implementing base classes that are common to all GPU profilers
+// inline in this header.
+
+class ProfilerActivityBuffer {
+ public:
+  ProfilerActivityBuffer()
+      : data_(nullptr), size_(0) {}
+
+  ProfilerActivityBuffer(const char* data, size_t size)
+      : data_(std::make_unique<char[]>(size)), size_(size) {
+    memcpy(data_.get(), data, size_);
+  }
+
+  ProfilerActivityBuffer(const ProfilerActivityBuffer& other)
+      : ProfilerActivityBuffer(other.GetData(), other.GetSize()) {}
+
+  ProfilerActivityBuffer(ProfilerActivityBuffer&& other)
+      : ProfilerActivityBuffer() {
+    std::swap(data_, other.data_);
+    std::swap(size_, other.size_);
+  }
+
+  ProfilerActivityBuffer& operator=(const ProfilerActivityBuffer& other) {
+    if (&other == this) {
+      return *this;
+    }
+
+    new (this) ProfilerActivityBuffer{other};
+    return *this;
+  }
+
+  ProfilerActivityBuffer& operator=(ProfilerActivityBuffer&& other) {
+    if (&other == this) {
+      return *this;
+    }
+
+    new (this) ProfilerActivityBuffer{std::move(other)};
+    return *this;
+  }
+
+  static ProfilerActivityBuffer CreateFromPreallocatedBuffer(char* data, size_t size) {
+    ProfilerActivityBuffer res{};
+    res.data_.reset(data);
+    res.size_ = size;
+    return res;
+  }
```

Comment:
This makes the owership a bit convoluted. Because you can not figure out how the `char* data` is mean to be maintained from the function signature. How about make it `std::unique_ptr<char[]>&&` and let user to construct it.

Note, this is coupled with the comment in `CUPTIManager::BufferRequested`.

Reply:
Thanks for the suggestion: it's definitely cleaner :)